### DISCUSSION TITLE
Command-line example for Helidon MP?

### DISCUSSION THREAD

**Discussion by lprimak** [helidon-io/helidon#8072] [Q&A] [answered]
Hi, I am trying to write a command-line app in Helidon MP using CDI.
No web server, jax-rs or alike is required.

However, I cannot find any example that does this. Every example is about running rest calls, or other MP features which I do not need for this, just CDI is a requirement.

My original application uses Weld SE, but I would like to try out native image with Helidon.

Thank you.

**Comment by lprimak** [comment]
Anybody? @danielkec @ljnelson my CDI fella? Please?

**Reply by ljnelson** [reply]
Have some personal things going on but what you are basically asking for is still essentially Weld SE, but the amazing repackaging we did of it to shoehorn it into native image. I’ll let others comment but I would gently suggest that native image isn’t All That in this case.

**Reply by lprimak** [reply]
Thanks @ljnelson !
I have indeed noticed, and already attempted to reuse this "amazing repackaging" but something is still missing with Weld proxies and it's failing.

My project is in
https://github.com/flowlogix/native-java/ (I have Helidon, Quarkus and Weld-SE version)
Quarkus works, Helidon is bringing in web / rest server (which I do not want), and doesn't take
command-line arguments into account, and Weld-SE doesn't work due to some proxy issues :(

Indeed I agree native image "isn't all that" but for command line distributions, I think it's the best I can do for now :)

**Reply by lprimak** [reply]
Is there anything else "special" about how Helidon does Weld SE besides changing the proxy names? Am I missing something else? Or am I "just doing something wrong?"

**Reply by lprimak** [reply]
Ok, just found helidon-mp-graal-native-image-extension that seems relevant. I will try using that tomorrow or the next day. Hopefully that's the key missing piece. If there is something else I am missing please tell me :)

Thank you!

**Reply by tomas-langer** [reply]
Helidon provides a library `helidon-microprofile-cdi` that is only "pure" CDI + configuration. If you base your application on this, you will not get any JAX-RS or server side features.
Also if you use our project template (and just reduce dependencies to this module), you can use our native image build using `mvn package -Pnative-image` (it will indeed use our extension you have discovered)

**Reply by lprimak** [reply]
Do you know offhand it that will allow direct Weld SE initialization?  As the other issue is that I am unable get to command line arguments

**Reply by danielkec** [reply]
Hi Lenny, we did small example just for you:
[cli-mp-helidon.zip](https://github.com/helidon-io/helidon/files/13488988/cli-mp-helidon.zip)

**Reply by danielkec** [reply]
Easiest way how to sent cli args is thru config, sys props are the default config source so something like this:
```java
    public static void main(String[] args) throws IOException {
        System.setProperty("cli.args", String.join(",", args));
...
```
could help
```java
    @Inject
    @ConfigProperty(name = "cli.args", defaultValue = "")
    List<String> args;
```

**Reply by ljnelson** [reply]
Note also: https://stackoverflow.com/a/67117216/208288 Not sure if we honor that when we did our repackaging etc.

**Reply by lprimak** [reply]
The `helidon-microprofile-cdi` +  weld repackaging + `helidon-mp-graal-native-image-extension` did the trick.

Just wondering, there a way to initialize Weld SE directly without using `helidon-microprofile-cdi`? Or using some part of it?
The issue is that I want to use Weld version that's not the same as the one Helidon's "amazing repackaging" provides.

I figured out how to produce the same repackaging for different versions of Weld. However,  am missing some magic that `helidon-microprofile-cdi` provides still to initialize Weld SE directly.

Thank you.

**Reply by danielkec** [reply]
:eyes: @tomas-langer

**Reply by lprimak** [reply, marked as answer]
Alright, I found they "key ingredient" Weld needs to be statically initialized at build-time (native image):
```
    static {
        container = new Weld()
                .addServices(new HelidonProxyServices())
                .initialize();
    }
```

This way, no runtime initialization is performed and it works!

**Comment by lprimak** [comment]
Further looking at the Weld `ProxyFactory` code in Helidon, could this be done cleaner via SubstrateVM Substitution mechanism by any chance?

**Reply by tomas-langer** [reply]
Problem with Substitutions in GraalVM is that they are not that much promoted by the team and are considered a "hackish" solution. It is always preferable to do it differently (ideally by modifying the original library).

**Comment by lprimak** [comment]
I have created https://github.com/flowlogix/weld-native which uses GraalVM substitutions and ByteBuddy for non-GraalVM situations to patch Weld the same way Helidon does, but without modifying any code directly or shaded version of Weld.

**Reply by lprimak** [reply]
What do you think @tomas-langer ?

**Reply by lprimak** [reply]
Hi, @tomas-langer @ljnelson 
I have renamed the project to https://github.com/flowlogix/weld-native
What do you think? I would like to contribute this to Helidon, IMHO its cleaner than the current way Helidon patches Weld.

Thank you!

**Reply by lprimak** [reply]
Usage example is here: https://github.com/flowlogix/native-java/tree/main/helidon/helidon-greeter

**Reply by tomas-langer** [reply]
Great job! I know how complicated it can be to weld Weld and `native-image` together (pun intended).
This solution is not really an option for us in Helidon, as one of the main rules we stand by (in our codebase) is not to play tricks on the classes - so ByteBuddy is out of the question. Anytime bytecode manipulation happens, we get something different than the sources tell us, which makes troubleshooting much harder. Also this requires some assumptions to be made about the classpath that I prefer not to make

**Reply by lprimak** [reply]
Maybe you can reconsider since time has passed?
I now see many GraalVM substitutions in Helidon itself, which does exactly the same thing as ByteBuddy does for running Java code.

The only reason even to have ByteBuddy is to have the same functionality on the JVM as GraalVM substitution does in native image.
